src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -641,11 +641,12 @@ Deoptimization::UnrollBlock* Deoptimization::fetch_unroll_info_helper(JavaThread
   bool caller_was_method_handle = false;
   if (deopt_sender.is_interpreted_frame()) {
     methodHandle method(current, deopt_sender.interpreter_frame_method());
-    Bytecode_invoke cur = Bytecode_invoke_check(method, deopt_sender.interpreter_frame_bci());
-    if (cur.is_invokedynamic() || cur.is_invokehandle()) {
-      // Method handle invokes may involve fairly arbitrary chains of
-      // calls so it's impossible to know how much actual space the
-      // caller has for locals.
+    Bytecode_invoke cur(method, deopt_sender.interpreter_frame_bci());
+    if (!cur.is_invokedynamic() && MethodHandles::has_member_arg(cur.klass(), cur.name())) {
+      // This should cover all real-world cases.  One exception is a pathological chain of
+      // MH.linkToXXX() linker calls, which only trusted code could do anyway.  To handle that case, we
+      // would need to get the size from the resolved method entry.  Another exception would
+      // be an invokedynamic with an adapter that is really a MethodHandle linker.
       caller_was_method_handle = true;
```

Comment:
This flag also controls the code at 711 that controls the computation of caller_adjustment.  Is the new answer also correct for that code?

This code might be a bit clearer if the computations of  caller_was_method_handle, caller_adjustment and the new caller_actual_parameters were all closer together, though that might complicate a backport so maybe it should be deferred to some later cleanup.

Reply:
Yes, I have further cleanup that I want to do later, but I want to minimize changes in this one to simplify backports.
Good catch about line 711.  I left it in on purpose, again to simplify backports, but it could be safely removed.  All it does here is over-estimate the adjustment, which is harmless.  In future cleanups, I hope to make the adjustment exact rather than a possibly over-estimated increment.

Reply:
Sounds good.  I kind of assumed it was a benign oversizing.